lib/resourcemerge/core_test.go

```diff
@@ -262,3 +262,284 @@ func TestEnsurePodSpec(t *testing.T) {
+					},
+				},
+			},
+			input: corev1.Service{
+				Spec: corev1.ServiceSpec{},
+			},
+			expectedModified: true,
+			expected: corev1.Service{
+				Spec: corev1.ServiceSpec{},
+			},
+		},
+		{
+			name: "replace port (same port name, different port)",
+			existing: corev1.Service{
+				Spec: corev1.ServiceSpec{
+					Ports: []corev1.ServicePort{
+						{
+							Name: "test",
+							Port: 8080,
+						},
+					},
+				},
+			},
+			input: corev1.Service{
+				Spec: corev1.ServiceSpec{
+					Ports: []corev1.ServicePort{
+						{
+							Name: "test",
+							Port: 8282,
+						},
+					},
+				},
+			},
+			expectedModified: true,
+			expected: corev1.Service{
+				Spec: corev1.ServiceSpec{
+					Ports: []corev1.ServicePort{
+						{
+							Name: "test",
+							Port: 8282,
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "replace port (no name, different port)",
+			existing: corev1.Service{
+				Spec: corev1.ServiceSpec{
+					Ports: []corev1.ServicePort{
+						{
+							Port: 8080,
+						},
+					},
+				},
+			},
+			input: corev1.Service{
+				Spec: corev1.ServiceSpec{
+					Ports: []corev1.ServicePort{
+						{
+							Port: 8282,
+						},
+					},
+				},
+			},
+			expectedModified: true,
+			expected: corev1.Service{
+				Spec: corev1.ServiceSpec{
+					Ports: []corev1.ServicePort{
+						{
+							Port: 8282,
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "replace multiple ports",
+			existing: corev1.Service{
+				Spec: corev1.ServiceSpec{
+					Ports: []corev1.ServicePort{
+						{
+							Name: "foo",
+							Port: 8080,
+						},
+						{
+							Name: "bar",
+							Port: 8081,
+						},
+					},
+				},
+			},
+			input: corev1.Service{
+				Spec: corev1.ServiceSpec{
+					Ports: []corev1.ServicePort{
+						{
+							Name: "foo",
+							Port: 8282,
+						},
+						{
+							Name: "bar",
+							Port: 8283,
+						},
+					},
+				},
+			},
+			expectedModified: true,
+			expected: corev1.Service{
+				Spec: corev1.ServiceSpec{
+					Ports: []corev1.ServicePort{
+						{
+							Name: "foo",
+							Port: 8282,
+						},
+						{
+							Name: "bar",
+							Port: 8283,
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "equal (same ports different other)",
+			existing: corev1.Service{
+				Spec: corev1.ServiceSpec{
+					Ports: []corev1.ServicePort{
+						{
+							Name: "foo",
+							Port: 8080,
+						},
+						{
+							Name: "bar",
+							Port: 8081,
+						},
+					},
+				},
+			},
+			input: corev1.Service{
+				Spec: corev1.ServiceSpec{
+					Ports: []corev1.ServicePort{
+						{
+							Name: "bar",
+							Port: 8081,
+						},
+						{
+							Name: "foo",
+							Port: 8080,
+						},
+					},
+				},
+			},
+			expectedModified: false,
+			expected: corev1.Service{
+				Spec: corev1.ServiceSpec{
+					Ports: []corev1.ServicePort{
+						{
+							Name: "foo",
+							Port: 8080,
+						},
+						{
+							Name: "bar",
+							Port: 8081,
+						},
+					},
+				},
+			},
+		},
```

Comment:
can you add a unit test wrt 

1) input service has `Protocol` set due to defaulting and required doesn't https://github.com/kubernetes/api/blob/11707872ac1cab52aefea0325f4698eaffeda3c8/core/v1/types.go#L3988-L3991

2)  `TargetPort` https://github.com/kubernetes/api/blob/11707872ac1cab52aefea0325f4698eaffeda3c8/core/v1/types.go#L4005

the API defaults to Port if unset, also it might be a string defaulted if the port is known IANA etc..

3) also `NodePort` https://github.com/kubernetes/api/blob/11707872ac1cab52aefea0325f4698eaffeda3c8/core/v1/types.go#L4013

as this value is assigned some value if none is already provided.

Reply:
@abhinavdahiya sorry still not clear what we want to cover. `EnsureServicePorts` as any other ensuring function in this library do not known about defaults. `EnsureServicePorts` is just mutating a struct when !equality.Semantic.DeepEqual and it will overwrite whatever `existing` has regardless this has been previously applied explicitly by a client or a by the api machinery defaulting.
The only think I can think of to increase coverage would be to apply the manifest after running `EnsureServicePorts`, then client.get the object back and validate that defaults are applied back properly, but the fake client does not apply any defaults unless we build a custom Reactor, therefore we'll be actually just covering our custom Reactor logic and this seems out of the scope to unit for `EnsureServicePorts`.
Let me know if this does not make sense to you and still want me to add anything.

Reply:
> EnsureServicePorts as any other ensuring function in this library do not known about defaults. EnsureServicePorts is just mutating a struct when !equality.Semantic.DeepEqual and it will clobber whatever existing has regardless this has been previously applied explicitly by a client or a by the api machinery defaulting.

I think we should prevent un-necessary updates, for the api machinery defaulting.. because that is known right...?
client changes are more than welcome to be over-written.

Reply:
> The only think I can think of to increase coverage would be to apply the manifest after running EnsureServicePorts, then client.get the object back and validate that defaults are applied back properly,

the unit tests already have an option of what exists in cluster currently and what we want..

for example, if the Protocol is empty (required), let's not run an update call because we know that it was defaulted. or when TargetPort is empty (required) make sure we are only running update when targetport-port differ because that is client change and not defaulting..

etc.. 

Even if we don't want to change these, unit-test cases that show that we are running update in presence of defaults is useful for future work when we can be smarter..

hopefully that provides more context...

Reply:
@abhinavdahiya thanks! Ok I see, I added defaulting for `Protocol` https://github.com/openshift/cluster-version-operator/pull/272/files#diff-f7675c7b39553e431cdc6c3ba8f5c482R244 and test coverage https://github.com/openshift/cluster-version-operator/pull/272/files#diff-415c13f11ffc32696c5d69b900b3fe58R551-R638
Other values might depend on the serviceType, I don't think we should redo api machinery defaulting logic here any farther.
Please let me know if it looks good now or still wanting any change